### PR TITLE
add tile type LOCK

### LINKED ISSUE #5: add numeric keypad
keypad to enter passcodes to lock/unlock doors , arm/disam security devices, ....

### PR DESCRIPTION
**Pull request by moinmoin-sh**
shows a key pad on a modal popup for entering a key

closes #5 

Added new tile type LOCK
shows a locked Lock icon if state = on
shows an unlocked Lock icon is state = off
This tile type can be used for the key pad, but every other tile type could be also.
The KeyPad popup has to be enabled in the cofig

grafik

Click on an KeyPad enabled tile opens a modal popup with key pad, back space and return button, input field and the icon of the calling tile.
Entered numbers are shown for 300 mSec and then replaced with a solid bullet.
Backspace deletes single character from the right.
Return sends the key entered as` stat/ `payload : 
` {"screen":1, "tile":1, "tiletype":"light", "type":"button", "event":"key", "state":"off", "keycode":"1234" }`
Return with empty input field closes the modal popup (escapes)

The key pad can be controlled with the following cmnd/ payload : ` '{"setlockstate":{"set":"STATE"}}'`
where STATE =
`"locked" `     -> sets the icon to locked, clears the input field
`"unlocked" ` -> sets the icon to unlocked, clears the input field
`"failed" `       ->sets the icon color to red
`"close" `       -> closes the modal overlay

There is no evaluation of the input in the FW. The visible behavior has to be controlled via mqtt by the backend.

grafik

### CODE REVIEW

**Comment by sumnerboy12** [comment]
Looks fantastic Moin! 

Just thinking about how this would work within my NodeRED flows... could we keep the `cmnd/` payload in the same format as the other tiles? Something like `{ "setstate": [ { "screen": 1, "tile": 1, "state": "locked" ] }`? 

And could the modal popup be self-contained so that clicking the tile launches it and then once the user hits `return` it closes the popup and sends the `stat/` message with the code entered. At that point the backend can check the code and send a `cmnd/` update with the tile state - either `locked`, `unlocked` or `failed`. 

I think this would be a much simpler flow to work with, rather than having to manage the popup state along with the tile state.

**Comment by moinmoin-sh** [comment]
Not sure if I explained the function clear enough. The STATE that I'm referring to is only valid for the key pad dialog. My intention is to leave the dialog open until the correct key was entered. This state is not fed back to the tile. The tile states are still `on` and `off`.
When the backend checks the key it can easily send also the `close `cmnd/ to end the dialog.

Sending the status can of course be changed to the format you suggest.

File: src/main.cpp [hunk 1]
```diff
@@ -109,8 +110,9 @@ const void *img3dPrint = &ios_3dprint_60;
 const void *imgRemote = &ios_remote_60;
 const void *imgLocked = &ios_locked_60;
 const void *imgUnLocked = &ios_unlocked_60;
+const void *imgCeilingFan = &ios_ceiling_fan_60;
 
-int _act_BackLight;
+                          int _act_BackLight;
```
[the remark this answers is not in the document]
Comment (moinmoin-sh) [reply src/main.cpp:115]: OCD
too late time of the day as usual

**Comment by moinmoin-sh** [comment]
I see your scenario. In my world the key and it's validation is one thing and the action resulting in entering the right key is the other thing. The kepad dialog can be closed imediately after the key is verified then the action is started and when that has finished the status of the actor is updated.
I belive you see the calling tile as the actor (mech key lock) . I understand the calling tile just as calling for a key for something like arming/diasarming an alarm. Never thought of a direct control of an actor. Does the lock icon of the tile makes sense in your  model or do we have different HW icons that call for a key, which would mean the key pad dialog should be handled as the drop down?

**Comment by plains203** [comment]
This current configuration would also allow you to “lock” the screen so to speak, so that without the correct pin you only see the pin pad. There could be a use case for this.

**Comment by sumnerboy12** [comment]
That is a good point @plains203.

**Comment by sumnerboy12** [comment]
@moinmoin-sh yes I see the tile as representing for example my front door lock - it displays the lock state. Clicking the tile allows me to change the state of that tile, but it is secured with a code. So perhaps your original plan to just use `on` and `off` states makes the sense. 

Then any tile type can be tagged as `secure` or similar - i.e. it is a normal button tile with an icon, text and sub-label, but when clicked launches the keypad popup.

From that point I accept your approach is probably the most flexible. The code is validated and once verified the popup is closed, which results in the `stat/` event for the standard button tile click.

This would mean you could easily disable the `secure` config for a given tile and it would revert to standard tile behaviour, without any changes needed in the backend flows.

Shall we discuss in discord?

**Comment by moinmoin-sh** [comment]
I think you described very well what I had in mind, thanks for putting it into a few clear sentences.

Yes, let's move the discussion to discord.

**Comment by moinmoin-sh** [comment]
the latest commit has the "any tile" concept integrated. 
Please see the updated description in the first post above.

**Comment by sumnerboy12** [comment]
Ok had a play with this, have a couple of comments;

1. Backspace when nothing entered should close the popup (same as hitting return with nothing entered)
2. Escaping out of keypad shouldn't generate a `"event": "change"` - you are escaping so I don't think it should generate any event, since nothing has happened or changed
3. Need to include the current tile type and state when the user enters a code and hits enter - otherwise we don't know what action/command to generate. So I wonder if you should keep the payload the same as a normal tile, but change to `"event": "key"` and add `"keycode": "1234".
4. I wonder if we should remove the icon from the keypad - it only makes sense if the tile type is "lock" - this would also simplify the `setlockstate` command options - I think you just need `success` and `failed` - `success` closes the popup, `failed` turns the code (dots) red and leaves the popup open

So in my mind the flow would be the following;

1. User clicks a tile configured with the key pad
2. Keypad is shown (with no indication of the current state)
3. User enters code and hits enter
4. A status event is generated `{"screen":1, "tile":1, "tiletype": "coffee", "type":"button", "event":"key" , "state":"off", "keycode": "1234" }` (example here is a key protecting the coffee machine)
6. If code is wrong, the backend sends `{ "setlockstate": { "set": "failed" } }` and the code/dots turn red and popup stays visible
7. If code is correct, the backend sends `{ "setlockstate": { "set": "success" } }` and the popup closes, and the backend actions that event (based on the current "state" of the tile) - i.e. sends the command to turn the coffee machine on and once it is on, generates a `cmnd/` message to update the tile state

Hope that is clear enough?

I think this way will be a lot simpler to manage, and should hopefully be a reasonable UX for users to follow.

**Comment by austinscreations** [comment]
this looks cool

**Comment by moinmoin-sh** [comment]
> Ok had a play with this, have a couple of comments;
> 
>     1. Backspace when nothing entered should close the popup (same as hitting return with nothing entered)
> 
Sorry, but I don't like that TBH. That's not common behavior I think. I just hit the backspace until the entry field is clear, not counting the number of presses. Could easily be one too much and the popup closes unintended

>     2. Escaping out of keypad shouldn't generate a `"event": "change"` - you are escaping so I don't think it should generate any event, since nothing has happened or changed
> 
Yes, that's how I wanted it. Must have overlooked it ...

>     3. Need to include the current tile type and state when the user enters a code and hits enter - otherwise we don't know what action/command to generate. So I wonder if you should keep the payload the same as a normal tile, but change to `"event": "key"` and add `"keycode": "1234".
> 
OK, that makes sense.

>     4. I wonder if we should remove the icon from the keypad - it only makes sense if the tile type is "lock" - this would also simplify the `setlockstate` command options - I think you just need `success` and `failed` - `success` closes the popup, `failed` turns the code (dots) red and leaves the popup open
> 
There are two different use cases I think.
a) use it to arm/disarm an alarm, or lock/unlock a door (my thinking)
b) use it as access control (kids protection) for a switch (your thinking)

For
 a) the lock icon makes sense, because it shows the recent state. The final state could be shown after the key is validated - not required, just an option
 b) it doesn'n make sense, because the state is allways `locked` and entering the key is allways required. For b) it could make sense to show the icon of the calling tile as a reminder for the user to see at least what tile type caused the popup.

> 
> So in my mind the flow would be the following;
> 
for case b)

>     1. User clicks a tile configured with the key pad
> 
>     2. Keypad is shown (with no indication of the current state)
> 
but the icon of the calling tile

>     3. User enters code and hits enter
> 
>     4. A status event is generated `{"screen":1, "tile":1, "tiletype": "coffee", "type":"button", "event":"key" , "state":"off", "keycode": "1234" }` (example here is a key protecting the coffee machine)
> 
>     5. If code is wrong, the backend sends `{ "setlockstate": { "set": "failed" } }` and the code/dots turn red and popup stays visible
> 
the icon would turn red

>     6. If code is correct, the backend sends `{ "setlockstate": { "set": "success" } }` and the popup closes, and the backend actions that event (based on the current "state" of the tile) - i.e. sends the command to turn the coffee machine on and once it is on, generates a `cmnd/` message to update the tile state
> 
`{ "setlockstate": { "set": "close" } }`  ->  `{ "setlockstate": { "set": "success" } }` 
`"close"` does allready what you want.

The other `setlockstate ` cmnd/s are optional, you don't need to use them for this use case

> 
> Hope that is clear enough?
> 
> I think this way will be a lot simpler to manage, and should hopefully be a reasonable UX for users to follow.

Hope the commented comments aren't confusing

**Comment by sumnerboy12** [comment]
Ok - I figured you would say that about the backspace. Personally I found I expected the popup to close when it was empty and I hit backspace. But I am ok with your point of view on that one.

I was going to suggest showing the tile icon in the keypad popup but didn't want to push my luck! If you can do that then great!!

I understood that `close` === `success` but I was trying to suggest a format that was easier for users to follow and understand. I think a simple API makes life easier for all, and I don't see a use-case for updating the icon upon entering a valid code - that would result in the popup closing immediately so you are never going to see any icon change.

If we are now displaying the current tile icon, then I think the `locked` and `unlocked` states are redundant anyway?

File: src/main.cpp [hunk 2]
```diff
@@ -227,16 +227,17 @@ void publishDropDownEvent(classTile *tPtr, int listIndex)
...
 void publishKeyPadEvent(classTile *tPtr, const char *key)
 {
   StaticJsonDocument<128> json;
   json["screen"] = tPtr->getScreenIdx();
   json["tile"] = tPtr->getTileIdx();
   json["tiletype"] = tPtr->getTypeStr();
-  json["type"] = "key";
-  json["event"] = "change";
-  json["state"] = key;
+  json["type"] = "button";
```
Comment (sumnerboy12) [inline src/main.cpp:237]: Shouldn't this be from the tile config? I.e. what if you have a key on a dropdown or level tile?

Comment (moinmoin-sh) [reply src/main.cpp:237]: key on dropdown is not supported. The tiles `LINK`, `DROPDOWN`, `REMOTE`, `KEYPAD ` ar all showing their popups/linkes tile on click on the tile and that can only be one of them. As of now Config can't handle the 1 out of 4.

Comment (sumnerboy12) [reply src/main.cpp:237]: Ok cool, makes sense

**Comment by moinmoin-sh** [comment]
> If we are now displaying the current tile icon, then I think the locked and unlocked states are redundant anyway?

`locked `and `unlocked `would make sense if the tile type is `LOCK ` (my thinking a) above) . I would like to leave them in, just  in case there is someone following my thoughts. As mentioned before - they are optional.

The reason why I called the lockstate "close" was that I think that's more universal and can be used to abort the popup as well and closing after success makes sense to me. 

I think we are pretty much in agreement. I'll make the changes as discussed later today (may be your tomorrow)

**Comment by sumnerboy12** [comment]
This is looking pretty good now Moin - the only thing I would request is that the code is cleared when the user enters an invalid code and the backend returns `failed`.

Currently if I enter an invalid code, I have to backspace out each digit before I can close the popup, which takes a while.

Otherwise seems to be working pretty sweet!

**Pull request merged**